hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,195 @@
+
+#ifndef MPCD_SPHERE_GEOMETRY_H_
+#define MPCD_SPHERE_GEOMETRY_H_
+
+#include <math.h>
+
+#include "BoundaryCondition.h"
+
+#include "hoomd/HOOMDMath.h"
+#include "hoomd/BoxDim.h"
+
+#ifdef NVCC
+#define HOSTDEVICE __host__ __device__ inline
+#else
+#define HOSTDEVICE inline __attribute__((always_inline))
+#include <string>
+#endif // NVCC
+
+namespace mpcd
+{
+namespace detail
+{
+//! Spherical droplet geometry
+/*!
+ * TODO: write proper description of what is being done in the class below
+ */
+class __attribute__((visibility("default"))) SphereGeometry
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param R confinement radius
+         * \param bc Boundary condition at the wall (slip or no-slip)
+         */
+        HOSTDEVICE SlitGeometry(Scalar R, boundary bc)
+            : m_R(R), m_bc(bc)
+            { }
+
+        //! Detect collision between the particle and the boundary
+        /*!
+         * \param pos Proposed particle position
+         * \param vel Proposed particle velocity
+         * \param dt Integration time remaining
+         *
+         * \returns True if a collision occurred, and false otherwise
+         *
+         * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel is updated
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the box, and try to avoid branching or absolute value calls. The sign used
+             * in calculations is 1 if the particle is out-of-bounds in the radial direction, 0 otherwise.
+             *
+             * We intentionally use > / < rather than >= / <= to make sure that spurious collisions do not get detected
+             * when a particle is reset to the boundary location. A particle landing exactly on the boundary from the bulk
+             * can be immediately reflected on the next streaming step, and so the motion is essentially equivalent up to
+             * an epsilon of difference in the channel width.
+             */
+            double r2 = pos.x*pos.x + pos.y*pos.y + pos.z*pos.z;
```

Comment:
You can use dot product to condense this. Also, you should use `Scalar` for floating-point numbers in HOOMD to avoid type conversion:
```suggestion
            Scalar r2 = dot(pos,pos);
```